rfc: event system (#6957)
tendermint/tendermint: docs/rfc/rc-005-event-system.rst

RFC 005: Event System
=====================

Changelog
---------

- 2021-09-17: Initial Draft (@tychoish)

Abstract
--------

The event system within Tendermint, which supports a lot of core
functionality, also represents a major infrastructural liability. As part of
our upcoming review of the RPC interfaces and our ongoing thoughts about
stability and performance, as well as the preparation for Tendermint 1.0, we
should revisit the design and implementation of the event system. This
document discusses both the current state of the system 

Background
----------

Current State of Events
~~~~~~~~~~~~~~~~~~~~~~~

The event system makes it possible for clients, both internal and external,
to receive notifications after a series of given events, such as new blocks,
new transactions, validator set changes, as well as intermediate events during
consensus. Because the event system is very cross cutting, the behavior and
performance of the event publication and subscription system has huge impacts
for all of Tendermint.

The subscription service is exposed over the RPC interface, but also powers
the indexing (e.g. to an external database,) and is the mechanism by which

> `BroadcastTxCommit` is able to wait for transactions to land in a block.

Reviewer: I think this is all mostly true, but there are a few things I think worth calling out, relevant to how the various components interact with each other:

1. I believe indexing doesn't require _subscribing_ to the event stream. I think events are written to the sinks as the node is generating and publishing the events, i.e., pre-pubsub.
2. Event subscription shares plumbing with the RPC service, but I believe is _only_ exported via the websocket endpoint, not gRPC or the HTTP (URL/JSON). That detail isn't crucial except to note that event subscription may be separable from the rest of the RPC machinery with some work (I'm working on a proposal for maybe doing that).
3. I believe the mempool manages the subscription used to wait for `BroadcastTxCommit` rather than the RPC service (the latter does wind up blocking on the result, but the streaming happens internally on this path).

Author: 1. Are indexing events buffered/dispatched independently, or could one of these systems put back pressure on the other. 
2. Sweet.
3. https://github.com/tendermint/tendermint/blob/master/rpc/core/mempool.go#L71-L86 it really looks like this happens in the rpc end point.

Reviewer: 1. It's not clear. The code makes it look like they are done independently, but the way the channels are set up, I think it's possible publishing _could_ block indexing. I'd have to trace it out, but I think the important bit is that it isn't expected to be dependent (except in order).
3. Ah yes, indeed it does.

Reviewer: So, I was wrong about my original point (1) above. The indexer _service_ works by subscribing to the event stream, so they are indeed connected. In fact, in principle the speed of _indexing_ could also slow down consensus, with the way it's hooked up right now (though in practice I think that's unlikely to happen). I misunderstood the dependency order.

In effect, this means the pubsub system is the spinal column of the entire consensus engine.  That clearly wasn't the intent, but the idea that the event bus is the primary mechanism for synchronization isn't entirely irrational if we are willing to design around that. The way things are wired right now it would be a little tricky to decouple them, but I think it's doable.

That said, I'm becoming more and more convinced the event system needs to be given more weight in the design space, for all the reasons we've discussed here and elsewhere.

> The current pubsub mechanism relies on a couple of buffered channels,

Reviewer: Specifically: The subscriber engine maintains one as an inbound work queue, and there is another channel per subscriber-client (which may or may not be buffered) for publication, plus at least one other unbuffered(?) channel for waiting on shutdown. A slow subscriber is a DoS on the engine's work loop, which doesn't block new subscribers from being queued, but does prevent them from being registered, and can block publication. It's not quite deadlock, but it may be molasses-lock 🙂

Author: Yuuuuup. 

I'm going to leave this comment here for posterity.

primarily between all event creators and subscribers, but also for each
subscription. The result of this design is that, in some situations with the
right collection of slow subscription consumers the event system can put
backpressure on the consensus state machine and message gossiping in the
network, thereby causing nodes to lag.

Improvements
~~~~~~~~~~~~

> The current system relies on implicit queues built by the buffered channels,

Reviewer: ```suggestion
The current system relies on implicit, bounded queues built by the buffered channels,
```

and though threadsafe, can force all activity within Tendermint to serialize,

> which does not need to happen. Additionally, timeouts for subscription

Reviewer: We do probably need to make some guarantees about order of events—if clients need the event stream to reflect the partial order of what's happened in consensus, we probably can't let the consensus state machine and the event stream get too far apart. So, for example, maybe we do have to stall if indexing is slow. But I very much agree that a slow subscriber shouldn't be able to stall consensus—and we should probably define some service level objectives for things like:

1. How fast you need to be able to consume messages.
2. What happens to messages if you don't go fast enough (drop messages? disconnect subscriber? maybe some combination depending on how slow things get?)
3. How much history you can expect to catch up on if you lose your connection to the stream, and how to cursor into the stream

Some of these things are kind of implementation defined right now.

Author: My (very rough) conjecture is that we don't actually need to publish very many events from consensus, and I think winnowing the types of events before we start implementing QoS features for the event system will be useful, but I think these are good ways to frame the requirements and objectives.

consumers related to the implementation of the RPC layer, may complicate the
use of the system.

References
~~~~~~~~~~

- Legacy Implementation
  - `publication of events <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L333-L345>`_ 
  - `send operation <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L489-L527>`_ 
  - `send loop <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L381-L402>`_
- Related RFCs 
  - `RFC 002: IPC Ecosystem <./rfc-002-ipc-ecosystem.md>`_ 
  - `RFC 003: Performance Questions <./rfc-003-performance-questions.md>`_ 

Discussion
----------

Changes to Published Events
~~~~~~~~~~~~~~~~~~~~~~~~~~~

As part of this process, the Tendermint team should do a study of the existing
event types and ensure that there are viable production use cases for
subscriptions to all event types. Instinctively it seems plausible that some
of the events may not be useable outside of tendermint, (e.g. ``TimeoutWait``

> or ``NewRoundStep``) and it might make sense to remove them. Certainly, it

Reviewer: Remove, or at least filter out by default.

Reviewer 2: We likely can remove, but it's worth noting that the `internal/consensus/` tests rely on these events.

would be good to make sure that we don't maintain infrastructure for unused or

> un-useful message indefinitely.

Reviewer: Strong agree. And possibly some things are implicitly not indexed? Event attributes currently have a flag for whether they should be indexed, but events as a whole do not.

Reviewer 3: > And possibly some things are implicitly not indexed? 

Well none of the events (that the pubsub system supports) are indexed with the one exception of `EventDataNewBlockHeader` which you could argue should be a separate struct for that exact reason. We want to make a clear distinction between the events in the pubsub system and the application events that are emitted from begin block, end block and after every transaction and that use the pubsub as the vehicle to support the `EventSink`

Blocking Subscription
~~~~~~~~~~~~~~~~~~~~~

The blocking subscription mechanism makes it possible to have *send*
operations into the subscription channel be un-buffered (the event processing
channel is still buffered.) In the blocking case, events from one subscription
can block processing that event for other non-blocking subscriptions. The main
case, it seems for blocking subscriptions is ensuring that a transaction has

> been committed to a block for ``BroadcastTxCommit``. Removing blocking

Reviewer: If BroadcastTxCommit is really the only use case for "blocking" subscriptions, then we can definitely find a better solution (even a webhook would probably be better than freezing consensus). But if there are other use cases, this is another case where we should define the service levels more explicitly.

I'm also not convinced BroadcastTxCommit is a service we should keep baked into the consensus node. If we have a way for a client to subscribe and catch up on an event stream reliably, then the waiting logic can be moved out of the core into the client (or a library the client calls). Broadcast a transaction in the usual way, then go do other things while you wait for the appropriate event to fire. (We might need to support better filtering. Possibly a proxy would make sense, assuming only certain nodes need to handle subscriber traffic)

Reviewer: Yes, that might be sufficient for the common case. It might be worth doing this as a UX affordance.

That said: I think we have to deal with the more general issue of how a client discovers when/if its Tx lands in the chain, even if a connection fails or it takes longer than 10 seconds. My intuition is that BroadcastTxCommit should probably not exist, or should be built on top of a more reliable mechanism (e.g., inject the transaction and {monitor events, watch a light client, other}) and exposed as library functionality. I do _not_ think it belongs in the primary consensus API.

To really solve that we'll need input from users about how they're using this now, and how it does(n't) serve their needs. It's possible it might not implicate the event system at all, though I suspect it probably will.

subscriptions entirely, and potentially finding another way to implement

> ``BroadcastTxCommit``, could lead to important simplifications and

Reviewer 2: My general instinct here is that `BroadcastTxCommit` is unlikely to be all that useful as is. Since Txs can be low priority and therefore never make it into a block, or wait for a very long time, clients are very likely to timeout waiting for a response in that scenario. 

A simplification would likely put the onus on clients to poll the state to see if their Tx was included in a block or booted from the mempool.

Reviewer 2: Looks like the cosmos SDK has noted this problem:
https://github.com/cosmos/cosmos-sdk/blob/260c5728ab3b759788be68f04dd5d0ec07be0b78/client/broadcast.go#L88

Reviewer: Yes, agreed that this change would require a lot of coordination. The part I think is worth memorializing in the RFC is the desire to move in that direction (about which we seem more-or-less in accord).

Reviewer 3: I also concur with this. The events between getting a transaction in the mempool and having it committed in a block are distinct enough that although users may tend to want a "fire and wait a reasonable amount of time for it to be confirmed", I think it makes sense for Tendermint to have them as separate actions (submit a transaction, poll for confirmation) and rely on the front-end client to do the lifting. If we are going down this route, I think there are few adjustments we should consider making to improve the visibility of the mempool (some of these points stem from the conversation we had at the office hours):
1. RPC clients should be made aware of the TTL of a transaction to know the upper bound with which they can expect a transaction to be delivered by before having to retry. This is however a little bit more complicated than what it may seem because the TTL settings are node specific - another node could hold that transaction for eternity and propose it in a block after that time. At the end of the day it's up to the application (in most cases the SDK) to handle retry and concurrent transaction logic. We can just do the best we can to support that.
2. RPC clients could be made aware of what the minimum priority that was required to put a transaction in the last block was. AFAIK, there's no way for clients to know if the priority of their submitted transaction is anywhere near what the priority of other transactions are. 
3. RPC clients could be made aware of what transactions were dropped. This would be helpful so that clients know they need to resend their transactions but once again this is a node-specific element and some nodes could support enormous mempools in which nothing gets dropped. There are some circumstances where the dropping of a transaction is system-wide which happens when transactions get re-check-tx'ed.

Reviewer 2: >RPC clients should be made aware of the TTL of a transaction to know the upper bound with which they can expect a transaction to be delivered by before having to retry. This is however a little bit more complicated than what it may seem because the TTL settings are node specific - another node could hold that transaction for eternity and propose it in a block after that time

This definitely adds a bit of complexity since the guarantee we can make is a bit weaker. I do think it may be worth considering changing our mempool TTL to be a consensus param and enforcing it network-wide. Tx could then be submitted with a `HeightSubmittedAt` which would be gossiped along with the Tx so nodes would all know when to evict it. 

Worth considering these options if we do:

https://github.com/tendermint/tendermint/issues/6958

improvements to throughput without requiring large changes.

Subscription Identification
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Before `#6386 <https://github.com/tendermint/tendermint/pull/6386>`_, all
subscriptions were identified by the combination of a client ID and a query,

> and with that change, all it became possible to identify subscription's based
> on an ID but compatibility with the legacy identification means that there's a
> good deal of legacy code as well as client side efficiency that could be
> improved. 

Reviewer: ```suggestion
and with that change, it became possible to identify all subscription given
only an ID, but compatibility with the legacy identification means that there's a
good deal of legacy code as well as client side efficiency that could be
improved. 
```

(I think?)

Pubsub Changes
~~~~~~~~~~~~~~

> The pubsub core should be implemented in a way that removes the possibility of

Reviewer: Strong +1 for this high-level idea. I've got some thoughts about how to structure it so we can replace what we're doing now without too much disruption. I can put together a prototype we can play with.

Reviewer: I will need some time to write more detail, but here is a high-level synopsis:

1. Split the event system into two layers. 
2. The "inner" layer is a pure time-series with no filtering. Events can only be published to the inner layer, and any component of the node can publish. The inner stream has a (probably configurable) fixed buffer capacity and publishers block when that capacity is reached (see (3)).
3. The "outer" layer is the (only†) obligate subscriber to the inner layer. Its primary job is to drain the inner time-series and index the results. Indexing is the only operation that can push back on the publishers. In addition, the outer layer manages any/all "external" subscribers. A node that doesn't want to expose events can use a null event sink and disable subscriptions from outside the node, and everything will still work.
4. Each "external" subscriber maintains a filter (a function) and a fixed-capacity buffer. When the outer layer delivers an event to an external subscriber, it does not block: The event is either buffered for delivery, or discarded (whether it discards some other event and buffers the new one, or drops the new one, is a separate policy question, but those are basically the two visible outcomes).
5. Each event delivered to the client comes with a "cursor", notionally a timestamp or an index into the underlying time series. This allows a client to specify a point at/after which it wants to "catch up" on events. The amount of event history the node is willing to replay becomes a configurable setting, so clients have enough time to catch up but cannot force the node to maintain unbounded storage.
6. Replaying "old" event state can be handled independently from the event bus: When a client shows up wanting to subscribe to a cursor that is "old", the service can send back (batches of) old event data until it "catches up" then hand off the subscription to the outer layer of the event system to keep going on the "normal" path. This means registration would be a little more complex than it is now, but only for users who care about history.

All of this is very informal right now, but I think it's basically sound, and I think we could do most of the internal plumbing without breaking the existing APIs at first.

† In principle we might want other internal components of the node to be able to subscribe to the inner time series. That would also be fine, and it wouldn't break the model, but I can't think of any important use for that yet.

Reviewer: I meant to add: Part of the motivation for decoupling registration/replay from "running" subscribers is to allow us to shed load in a principled way: At the moment, a large number of subscribers can put more or less unlimited memory pressure on the node, even if they are timely about consuming their feed. The node should be able to tell a client at the point of registration "I don't have room for any more subscribers", so it can control both memory usage and concurrency in a predictable way. If there is a need for more capacity, an operator can increase capacity by running one or more "proxy" subscribers that mux external subscription clients onto a smaller set of node subscriptions. We could even write such a proxy—it would need to do some work to compile filters together and so on, but a lot of the code could be shared. The net result is that the system _can_ be scaled up, but the consensus node still has predictable resource footprint and stability characteristics.

Author: I think this is more or less what I was thinking of, and think your description is pretty good. I presume that the inner "timeseries" data is readable my subscribers (via their cursors,) after passing through the filter, and only *writeable* by the incoming event stream. 

My inclination is that we should implement this on top of the database layer (e.g. put it on disk rather than memory, and let the storage engine's concurrency control manage most of this,) the "hard" part then becomes deleting old events in an efficient way, as MVCC systems tend to be pretty bad at deletion, but that turns a big collection of hard problems into just one moderately gnarly problem.

Reviewer: > I presume that the inner "timeseries" data is readable my subscribers (via their cursors,) after passing through the filter, and only _writeable_ by the incoming event stream.

That's the general idea, yes. The exact point of division between the "inner" and "outer" layers I described isn't crucial—you could imagine, for example, making the "inner" layer responsible for indexing. I described it this way so that there are as few configuration options as possible grafted to the spinal column of the consensus node (e.g., indexing may be disabled entirely without loss of function for the inner layer).

> My inclination is that we should implement this on top of the database layer (e.g. put it on disk rather than memory, and let the storage engine's concurrency control manage most of this,) the "hard" part then becomes deleting old events in an efficient way, as MVCC systems tend to be pretty bad at deletion, but that turns a big collection of hard problems into just one moderately gnarly problem.

Yeah, that's more or less what I was thinking about for indexing as well. The "in memory" part is really just a working buffer. By allowing the indexer to push back the publishers, we can control how much stuff has to land in persistent storage before we "move on". I'm sure we would need to refine the details, but ideally I want to reuse as much of the other infrastructure (e.g., the database) as possible.

backpressure from the event system to impact the core system *or* for one
subscription to impact the behavior of another area of the
system. Additionally, because the current system is implemented entirely in
terms of a collection of buffered channels, the event system (and large
numbers of subscriptions) can be a source of memory pressure. 

These changes could include: 

- explicit cancellation and timeouts promulgated from callers (e.g. RPC end
  points, etc,) this should be done using contexts.

- subscription system should be able to spill to disk to avoid putting memory
  pressure on the core behavior of the node (consensus, gossip).
  
- subscriptions implemented as cursors rather than channels, with either

>   condition variables to simulate the existing "push" API or a client side

Reviewer 2: I'd be a little hesitant to perform to much of a large scale rebuild of the internal system until we have a better understanding of the throughput. At some level, we have a system that behaves like a big queue with multiple readers. I definitely think we should avoid letting readers starve each other out in the way we have it now, but I also think that if you have a big queue and your throughput isn't good enough, eventually you're going to back up. I similarly worry about this with the spilling to disk case. If we have to start spilling to disk because a reader is too slow, maybe we don't ever stop spilling to disk?

Reviewer: I agree—this is part of what I was getting at above, about defining what capabilities we want the system to provide. I think a lot of these things could be improved without rewriting from scratch—the API isn't that complex, and there's a lot of low-hanging free improvements that wouldn't necessarily break other components. 

I _do_ think we should define the constraints, though, and use that to do larger changes. Event subscription seems to be important enough for a lot of users that IMO it's worth some investment.

Reviewer 3: > Event subscription seems to be important enough for a lot of users that IMO it's worth some investment.

The importance of event subscription has certainly evolved over the last year. I think prior to the introduction of relayers, the events was just a nice feature that some clients might have used via a websocket. It certainly wasn't designed with strong reliability/robustness guarantees in mind.

I tend to agree that the more immediate challenge is distilling what items can be resolved now (and in a non-breaking way) and what items need to be pushed back for a more large scale redesign
